libs/ml/include/ml/optimisation/optimiser.hpp

```diff
@@ -122,29 +121,59 @@ typename T::Type Optimiser<T, C>::Run(std::vector<ArrayType> const &data, ArrayT
   DataType loss_sum{0};
 
   SizeType step{0};
-  while (step < n_data)
+
+  // Prepare output data tensors
+  std::vector<ArrayType> batch_data;
```

Comment:
can batch_data and batch_labels be made private member variables? Under this current implementation we create a new std::vector for each of these once per epoch. Not the most expensive, but could be cheaper too